File: app/js/lib/algorithms/k_means.js
const KMeans = {
  k_means(data, k = 3, maxIterations = 100) {
    data = data.map(x => [x, 0]);
    let centroids = data.slice().sort(() => Math.random() - 0.5).slice(0, k).map(([x]) => [x, 0]);

    for (let iter = 0; iter < maxIterations; iter++) {
      const assignments = data.map(([x]) => {
        const closest = centroids.reduce((min, [c]) => {
          const dist = (x - c) ** 2;
          return dist < min.dist ? { centroid: c, dist } : min;
        }, { centroid: centroids[0][0], dist: Infinity }).centroid;
        return [x, closest];
      });

      const newCentroids = [...new Set(assignments.map(([_, c]) => c))].map(clusterId => {
        const points = assignments.filter(([_, c]) => c === clusterId).map(([x]) => x);
        const avg = points.reduce((a, b) => a + b, 0) / points.length;
        return [avg, clusterId];
      });

      if (JSON.stringify(newCentroids.map(([c]) => c)) === JSON.stringify(centroids.map(([c]) => c))) break;
      centroids = newCentroids;
    }
    return data.map(([x]) => [x, centroids.findIndex(([c]) => c === x) || 0]);
  }
};

module.exports = KMeans;

File: app/lib/algorithms/k_means.rb
# frozen_string_literal: true

module KMeans
  module_function

  def k_means(data, k = 3, max_iterations = 100)
    data = data.map { |x| [x.to_f, 0] }
    centroids = data.sample(k).map { |x,| [x, rand] }

    max_iterations.times do
      assignments = data.map do |x, _|
        [x, centroids.min_by { |c,| (x - c)**2 }.first]
      end

      new_centroids = assignments.group_by(&:last).map do |cluster_id, points|
        avg = points.map(&:first).sum / points.size
        [avg, cluster_id]
      end

      break if new_centroids.map(&:first) == centroids.map(&:first)
      centroids = new_centroids
    end

    data.map { |x,| [x, centroids.find_index { |c,| c == x } || 0] }
  end
end
